pkg/kudoctl/cmd/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,176 @@
+package diagnostics
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/kudobuilder/kudo/pkg/kudoctl/env"
+	"github.com/kudobuilder/kudo/pkg/version"
+
+	"github.com/spf13/afero"
+)
+
+type Options struct {
+	Instance string
+	LogSince int64
+}
+
+func NewOptions(instance string, logSince time.Duration) *Options {
+	opts := Options{Instance: instance}
+	if logSince > 0 {
+		sec := int64(logSince.Round(time.Second).Seconds())
+		opts.LogSince = sec
+	}
+	return &opts
+}
+
+func Collect(fs afero.Fs, options *Options, s *env.Settings) error {
+	ir, err := NewInstanceResources(options, s)
+	if err != nil {
+		return err
+	}
+	instanceDiagRunner := &Runner{}
```

Comment:
I like the `Runner` pattern here, as it abstracts away resource collection in a good way. What I don't like is the use of a mutating `processingContext`, because it makes to information flow harder to understand. We should decouple the collection of resources needed for the context from the collection of resources needed for the logs. It's okay if that means that some information is gathered twice. With a static context, the runner would become something like

```go
Run(&ResourceCollector{
			loadResourceFn: ir.Services,
			errKind:        "service",
			parentDir:      ctx.attachToInstance(),
			printer:        p,
			printMode:      RuntimeObject}
```
 while also removing the necessity for the `callback` parameter in `ResourceCollector`. This might not look like much, but is much simpler to understand.

Reply:
I assume @zen-dog would say some words here, as in my understanding this suggestion closely matches his concept of minimizing flexibility and generality.
I totally agree that _callback_ pattern modifying an object outside of the scope of the function is not the nicest of programming patterns. However, I chose it for conceptual reasons and here's how I see things:
- we have a **uniform** collection sequence definition, and uniformity is usually an easy concept. If we force some resource providers to differ and decide that, say, collecting pods is not the same as collecting services, and we in the same time want to avoid callbacks, then in order to expose the data previously extracted via callbacks we'd either have to break the collector interface or use the pod collector's internals.
- collector should preferably be **unaware** of how the information it collects is used otherwise than for printing and unaware of other collectors. Whatever we extract from collected objects, collector stays **the same**. The extraction is decoupled from collection. Note that collectors (and resource loaders) are also unaware of the context
- I haven't yet properly unexported what needs to be unexported. As long as this PR doesn't have the flexibility of that complicated one with plenty of indirections, `Run` should be unexported, and only whole sequences exported, and processing context should be local to `Runner`. I think having several packages here would be an overkill, but if we put Collectors it its own imaginary package exposing only the interface and constructor, and Runner in its own imaginary package exposing the sequences, things become clearer. Processing context is then managed in one place, and tracking its mutation is not more difficult than tracking a mutation of, say, local var in a function. The callback hell mutating the context from all over the code is not taking place.

Reply:
There a few things to unpack here but the gist of it is: while we still need a somewhat generic `ResourceCollector` which is a "working horse" of the collection, but we also need higher-level collectors that can utilize the lower ones.  Let's take an imaginary `PodCollector` as an example which ideally collects two files: pod definition and the logs like this:
```
├── pod_zk-zookeeper-0
│   ├── zk-zookeeper-0.log.gz
│   └── zk-zookeeper-0.yaml
```
All it needs to collect both are pod name and namespace. 

1. For the definition, it might construct a generic `ResourceCollector(ir.Pods, ...)` using a resource collection fund `ir.Pods`
2. For the logs, it can use a `LogCollector(...)`

The same goes for an imaginary `InstanceResourceCollector`. Currently, you have to invent the `ResourceCollectorGroup` for this case, but I'd rather have a specified collector as it is (subjectively) simpler to grok and you don't need the `ResourceCollectorGroup` abstraction. 

An imaginary `KUDOManagerCollector` can utilize all of the above to collect everything that is needed about the KUDO manager.

The **uniformity** should be in the fact that all collectors implement an interface like:
```
type Collector interface {
	func Collect(fs afero.Fs, ctx *Context,  kc kudo.Client s *env.Settings) error
}
```
and know how to collect the resources they need. 

P.S. I hope, that on the way, some of the interfaces become simpler. Like the fact that you needed to copy `type Object interface` from the k8s internals to generalize over all returned objects. Here too, I'd rather accept some code redundancy and opt-in into the typed resource collectors. **Premature over-generalization is the root of all evil.**

Reply:
I'd like for us to also think about dependencies. We **don't need** to implement the support for it now, but I'd very much like to keep the general architecture open enough to support it later. Just a brief reminder: with KEP-29 implemented, we could have instances owning (via `ownerReferences` field) other dependent instances e.g.:
```
AA
├── BB
│   ├── EE
│   └── GG
└── CC
```

I think it would help if instead of a `Runner` we would have a higher-level `InstanceCollector` which could (in the future) check the `ownerReferences` field and recursively instantiate itself to collect the dependency instances. 
So `instance-AA` collector instantiating `instance-BB`, which in turn instantiates `instance-EE` and so on. Each will take a `parentDir` via constructor so that the resulting file structure can look like:
```
diag
├── kudo
│   └── ...
├── operator_aa
│   ├── instance_aa
│   │   ├── operator_bb  // <-- nested operator BB files
│   │   │   ├── instance_bb
│   │   │   ├── pod_bb-0
│   │   │   │   ├── bb-0.log.gz
│   │   │   │   └── bb-0.yaml
│   │   │   ├── ....
│   │   │   └── bb.yaml
│   │   ├── pod_aa-0
│   │   │   ├── aa-0.log.gz
│   │   │   └── aa-0.yaml
│   │   └── aa.yaml
│   ├── operatorversion_aa-0.3.0
│   │   └── aa-0.3.0.yaml
│   └── aa.yaml
└── settings.yaml
```
Again, we don't have to tackle it now but it would be nice if the structure would allow dependencies with minimal overhead.

Reply:
Yes, I agree. Thought a bit more about this and how my approach could be done. Realized that your approach provides more readability by making resource collection more consistent.